### PR TITLE
Repo Guide

### PR DESCRIPTION
**Pull request by kmarsteller**
For makers of new repos that depend on OpenMDAO, some advice in the form of docs.
Also:
*coveragerc deemed unused, superseded by testflo cmds
*certain doc packages need to be exposed for import, e.g. embed directives
ONCE THIS GOES IN, I WILL NEED TO UPDATE DYMOS AND OAS immediately to change the imports in their conf.py files
*fixing several other small probs found in other docs

### CODE REVIEW

File: openmdao/docs/utils/generate_sourcedocs.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def generate_docs(dir, top, packages, project_name='openmdao'):
     if not os.path.isdir(packages_dir):
         os.mkdir(packages_dir)
 
-    # look for directories in the openmdao level, one up from docs
+    # look for directories in the openaero level, one up from docs
```
Comment (Kenneth-T-Moore) [inline openmdao/docs/utils/generate_sourcedocs.py]: openaero?

Comment (kmarsteller) [reply openmdao/docs/utils/generate_sourcedocs.py]: oh, whoops

**Comment by kmarsteller** [comment]
Looks like I didn't quite do the api.py imports right. (for instance I'm trying to import "embed_code" instead of the EmbedCodeDirective class...gotta fix that up.  So exclude that from the review for the moment while I figure that out in the morning.

File: openmdao/docs/developer_docs/repo_guide/github.rst [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+Personal Access Token
+---------------------
+
+A Personal Access Token is something created on Github, to allow another service (e.g. Travis) to access Github,
+with a certain amount of permission.  Once you create a Personal Access Token, it will be visible for
+one viewing, and you must copy it then use it immediately before you lose it (and then you have to make another one).
+
+To create a Personal Access Token:
+Log in to githb.com.  Go to Github *account* Settings (**not** Repo Settings)->Dev Settings->Personal Access Token
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/github.rst]: githb.com -> github.com

File: openmdao/docs/developer_docs/repo_guide/github.rst [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    images/AccessToken1.png
+
+Give the access token the name GH_TOKEN, and the permission "repo access," as pictured:
+
+.. image::
+    images/AccessToken2.png
+
+You'll need this token to allow the transfer of your documents to github-pages, once they are built on Travis CI.
+
+Then on Travis CI, you'll need to go to your Environment Variables section, and create a new env var, as shown here:
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/github.rst]: suggest: "env var" -> "variable" (since it's clear from context that you're talking about environment variables.)

File: openmdao/docs/developer_docs/repo_guide/github.rst [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    images/GithubPages.png
+
+travis-sphinx Usage
+-------------------
+
+`travis-sphinx` is a standalone script for automated building and deploying of sphinx docs to github-pages
+via TravisCI.
+
+Once your personal access token is set up, you can begin using travis-sphinx within your Sphinx configuration file. (`docs/conf.py`)
+The two calls that should be used are, in their respective doc sections:
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/github.rst]: I think remove "doc"?   -> "in their respective sections:"

File: openmdao/docs/developer_docs/index.rst [hunk 5]
```diff
@@ -45,3 +45,20 @@ in terms of formatting and embedding code.
...
+
+There are several things to take into account when setting up a repository that depends on
+OpenMDAO.  Github settings, getting documentation transferred to github-pages, setting up
+continuous integration on Travis CI, and releasing software are some of the topics covered.
+
+.. toctree::
+    :maxdepth: 1
+
+    repo_guide/docs.rst
+    repo_guide/github.rst
```
Comment (swryan) [inline openmdao/docs/developer_docs/index.rst]: looking at the github section, it seems like maybe it should come after the travis section (since you first have to know how to build stuff on travis, and it references section of travis.yml that is introduced there)

File: openmdao/docs/developer_docs/repo_guide/travis.rst [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+.. code-block:: python
+
+    script:
+        testflo -n 1 openmdao --coverage  --coverpkg openmdao --cover-omit \*tests/\*  --cover-omit \*devtools/\* --cover-omit \*test_suite/\* --cover-omit \*docs/\*;
+
+    after_success:
+        coveralls
+
+The point of the above example is to use testflo to run things, and set the `coverpkg` to your project, and the `cover-omit` dirs to exclude from coverage, use coveralls
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/travis.rst]: suggest: "run things" -> "run the test suite"
suggest rewording to explain that testflo generates coverage data (via the coverpkg and cover-omit options) and pointing to the testflo github for further information.  (maybe we should add a dump of `testflo --help` to the README there or make a quick wiki page showing the options)

File: openmdao/docs/developer_docs/repo_guide/release_process.rst [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+Once the release pull request from your release branch has been accepted, then another set of post-merge CI tests have passed,
+it’s time to “tag” the release. This must happen before any other commits hit master.
+Typing :code:`git tag` will show all the tags that exist up until now.
+
+:code:`git tag 1.0.0` will locally create a new tag, 1.0.0.  You can :code:`git tag` again to see that it exists now.
+For minor changes to the code, we will increment the third digit, e.g. 1.0.1
+For API changes to the code, we will increment the second digit, e.g. 1.1.0
+
+Now we need to get that tag up to master:  (Assuming you’re a repository owner for your project, and that `origin` is set to projectname/repo.)
+:code:`git push origin —-tags`
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/release_process.rst]: there is a long dash in "—-tags" instead of two normal dash marks

File: openmdao/docs/developer_docs/repo_guide/release_process.rst [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+When you do this, you should see the tags upload to the intended repository.
+
+You can also go to github.com website and look at OpenMDAO repo…in the same dropdown as branches, there’s a tab for tags also.  Make sure your new tag is there.
+
+pypi
+----
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/release_process.rst]: suggest an introductory sentence here about pypi and what you want to do (e.g. make you project installable via pip or whatever)

File: openmdao/docs/developer_docs/repo_guide/release_process.rst [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+OK, so how do we USE this to make a release?
+
+First, we need to build a source distribution, so from the top level, where setup.py lives.:
+
+:code:`python setup.py sdist`
+
+This will create a dist directory, in which lies `projectname-1.0.0.tar.gz`
+
+Finally, we upload this file up to pypi using `twine`:
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/release_process.rst]: "upload this file up to" -> "upload this file to"
also maybe make the twine reference a hyperlink to https://pypi.org/project/twine/?

File: openmdao/docs/developer_docs/repo_guide/release_process.rst [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+:code:`twine upload dist/projectname-1.0.0.tar.gz`
+
+You should be able to watch the dist upload. Then, go to your page at pypi.org and make sure the new release is visible there.
+
+Finally, tell the team and users that your release is done.  Wait for everything to fall apart.
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/release_process.rst]: 🤣

File: openmdao/docs/developer_docs/repo_guide/advanced_operations.rst [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+One of these steps, you will find out, is going to re-write your `.travis.yml` file and take out all the spacing,
+comments, and formatting. So if you like the way your `.travis.yml` file looks, I suggest backing up that file beforehand.
+You can then take the auto-generated lines that we’re about to generate, and stick them back in the pretty-looking `.travis.yml` file.
+
+To get this doc transfer automated, we need to be able to move built docs from Travis to our private server without any required input,
+because there is no human in the loop to type passwords or answer prompts.  This will require some ssh key wizardry.
+The overall concept is to have an encrypted key already on the Travis server, then to decrypt it, and use it to do a
+password-less transfer of the docs from Travis to our private server (in our case, Webfaction).  Here's the outline:
+
+    ::
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/advanced_operations.rst]: suggest a sphinx numbered or bulleted list here instead a literal block

File: openmdao/docs/developer_docs/repo_guide/advanced_operations.rst [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+**2. Encrypt the Key**
+    Again from the same directory, execute the next command to encrypt the PRIVATE key file that we just generated.
+    (NOTE: this is the aforementioned command that will change your .travis.yml file! ):
+
+    ::
+
+        travis encrypt-file travis_deploy_rsa --add -—debug --org
+
+        —-add adds the decryption command to the .travis.yml file automatically.
+        —-debug shows all the available output of the command.
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/advanced_operations.rst]: some long dashes here instead of two regular dashes

File: openmdao/docs/developer_docs/repo_guide/advanced_operations.rst [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+    B. Need to copy the public key generated above to our Webfaction server to allow passwordless entrance.
+
+     1. On webXXX, in the ~/.ssh folder, there is a file called authorized_keys.
+     2. Copy the contents of the travis_deploy_rsa.pub as an entry into the authorized_keys file.
+
+**4. Modify the YAML Further**
+
+   A. Late in the `before_install` section add this line:
+        `- echo -e "Host <server address>\n\tStrictHostKeyChecking no\n" >> ~/.ssh/config`
```
Comment (swryan) [inline openmdao/docs/developer_docs/repo_guide/advanced_operations.rst]: looks like you need to escape these backslashes for it to display correctly in the built page

File: openmdao/docs/developer_docs/repo_guide/travis.rst (hunk 6 above)
Comment (swryan) [reply openmdao/docs/developer_docs/repo_guide/travis.rst]: second thought.. add a project description (e.g. just `testflo --help`) to the testflo page on pypi and link to that

File: openmdao/docs/developer_docs/repo_guide/advanced_operations.rst (hunk 12 above)
Comment (kmarsteller) [reply openmdao/docs/developer_docs/repo_guide/advanced_operations.rst]: Not sure where those come from, but they're fixed.

File: openmdao/docs/developer_docs/repo_guide/docs.rst [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+~~~~~~~~~~~~~~~~~~~~~
+
+OpenMDAO's docs have a custom script that pre-processes all the .rst files found within a set of Sphinx documentation, and creates
+a custom blog-like tagging system that helps organize and cross-reference docs.
+
+The script finds occurrences of the .. tags:: directive and sets up the structure of the tags directory.  One file
+is created for each subject tag, and that file contains links to each instance of the tag throughout the docs.
+
+:code:`from openmdao.docutils import preprocess_tags`
+
```
[the remark this answers is not in the document]
Comment (kmarsteller) [reply openmdao/docs/developer_docs/repo_guide/docs.rst]: No, you're right, I just must have gotten distracted and not finished this thought.

File: openmdao/docs/developer_docs/repo_guide/github.rst [hunk 15]
```diff
@@ -0,0 +1,65 @@
+.. _`github_pages`:
+
+github-pages Repository Setup
+=============================
+
+If you want to publish to github-pages, from your docs built on Travis CI, make sure your Github repository is set up to be public,
```
[the remark this answers is not in the document]
Comment (kmarsteller) [reply openmdao/docs/developer_docs/repo_guide/github.rst]: I think I will re-word this and make the 'make your repo public' part into a warning directive

File: openmdao/docs/developer_docs/index.rst (hunk 5 above)
Comment (kmarsteller) [reply openmdao/docs/developer_docs/index.rst]: great feedback, thx

File: openmdao/docs/developer_docs/repo_guide/travis.rst (hunk 6 above)
Comment (kmarsteller) [reply openmdao/docs/developer_docs/repo_guide/travis.rst]: made a link and added a story to fix the testflo page on pypi
